### DISCUSSION TITLE
Teleport as SSH proxy - enable key exchange algorhythm?

### DISCUSSION THREAD

**Discussion by Kisuke-CZE** [gravitational/teleport#13883] [Q&A] [answered]
Hi,

when testing teleport, I run into another "problem".

I wanted to try to make our network device available via Teleport. From a teleport node, I can do ssh user@deviceIP:22.
But when I do same from teleport ssh proxy, I end up with:
```ssh: rejected: connect failed (ssh: handshake failed: ssh: no common algorithm for key exchange; client offered: [curve25519-sha256@libssh.org ecdh-sha2-nistp256 ecdh-sha2-nistp384 ecdh-sha2-nistp521], server offered: [diffie-hellman-group-exchange-sha256 diffie-hellman-group-exchange-sha1 diffie-hellman-group14-sha1 diffie-hellman-group1-sha1])```

Since I cannot do anything to adjust key-exchange methods on that network device (it's baked into FW and in real I am stuck what current version offers), I would like to enable `diffie-hellman-group-exchange-sha256` on the teleport side. Is it possible somehow?

**Comment by webvictim** [comment, marked as answer]
Search `kex_algos` under https://goteleport.com/docs/setup/reference/config/

I'm unsure whether you'll be able to add additional algorithms there, though - it's mainly intended to be able to take undesirable algorithms away rather than add new ones.

**Reply by Kisuke-CZE** [reply]
Thanks for a hint. It was the correct way to do this.

When I put `- diffie-hellman-group-exchange-sha256` to `teleport.yaml` I got an error:
```
ERROR: failed to parse Teleport configuration: KEX algorithm "diffie-hellman-group-exchange-sha256" is not supported; supported algorithms: ["curve25519-sha256@libssh.org" "ecdh-sha2-nistp256" "ecdh-sha2-nistp384" "ecdh-sha2-nistp521" "diffie-hellman-group14-sha1"]
```

But fortunately error message was pretty helpful (not every software has such a helpful error messages) and I was able to found compatible KEX algorithm: `- diffie-hellman-group14-sha1`

In the end, I was not able to get it working because device does not support something like `TrustedUserCAKeys`. It supports only SSH keys auth. But that's a problem outside of teleport.
